include/seqan3/alphabet/detail/alphabet_base.hpp

```diff
@@ -0,0 +1,237 @@
+
+#include <seqan3/alphabet/detail/member_exposure.hpp>
+#include <seqan3/core/detail/int_types.hpp>
+#include <seqan3/std/concepts>
+
+namespace seqan3
+{
+
+/*!\brief A CRTP-base that makes defining a custom alphabet easier.
+ * \tparam derived_type The CRTP parameter type.
+ * \tparam size         The size of the alphabet.
+ * \tparam char_t       The character type of the alphabet (set this to `void` when defining just a
+ *                      seqan3::semi_alphabet_concept).
+ *
+ * \details
+ *
+ * You can use this class to define your own alphabet, but types are not required to be based on it to model
+ * seqan3::alphabet_concept, it is purely a way to avoid code duplication.
+ *
+ * The base class represents the alphabet value as the rank and
+ * automatically deduces the rank type from the size, it further defines all required member functions and types; the
+ * derived type needs to define only the following two tables as static member variables:
+ *
+ *   * `static std::array<char_type, value_size> constexpr rank_to_char` that defines for every possible rank value
+ *     the corresponding char value.
+ *   * `static std::array<rank_type, 256> constexpr char_to_rank` that defines for every possible character value the
+ *     corresponding rank value.
+ *
+ * TODO add snippet
+ */
+template <typename derived_type, size_t size, typename char_t = char>
+class alphabet_base
+{
+public:
+    //TODO: this should be > 1 and there should be a different specialisation for size 1
+    static_assert(size > 0, "It does not make sense to use the base class for alphabets of size <= 1.");
+
+    /*!\name Member types
+     * \{
+     */
+    //!\brief The type of the alphabet when converted to char (e.g. via to_char()).
+    using char_type = char_t;
+    //!\brief The type of the alphabet when represented as a number (e.g. via to_rank()).
+    using rank_type = detail::min_viable_uint_t<size - 1>;
+    //!\}
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr alphabet_base() = default;
+    constexpr alphabet_base(alphabet_base const &) = default;
+    constexpr alphabet_base(alphabet_base &&) = default;
+    constexpr alphabet_base & operator=(alphabet_base const &) = default;
+    constexpr alphabet_base & operator=(alphabet_base &&) = default;
+    //!\}
+
+    /*!\name Read functions
+     * \{
+     */
+    /*!\brief Return the letter as a character of char_type.
+     *
+     * \details
+     *
+     * Satisfies the seqan3::alphabet_concept::to_char() requirement via the seqan3::to_char() wrapper.
+     *
+     * \par Complexity
+     *
+     * Constant.
+     *
+     * \par Exceptions
+     *
+     * Guaranteed not to throw.
+     */
+    constexpr char_type to_char() const noexcept
+    //!\cond
+        requires !std::Same<char_type, void>
+    //!\endcond
+    {
+        return derived_type::rank_to_char[rank];
+    }
+
+    /*!\brief Return the letter's numeric value or rank in the alphabet.
```

Comment:
`Return the letter's numeric value (rank) in the alphabet.` :nail_care: 
I understood it returns either or but it was meant as a synonym correct?